pkg/experiment/metastore/dlq/recovery.go

```diff
@@ -0,0 +1,152 @@
+package dlq
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	metastorev1 "github.com/grafana/pyroscope/api/gen/proto/go/metastore/v1"
+	"github.com/grafana/pyroscope/pkg/tenant"
+
+	objstore "github.com/thanos-io/objstore"
+	"io"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+type RecoveryConfig struct {
+	Period time.Duration
+}
+
+type LocalServer interface {
+	AddBlock(context.Context, *metastorev1.AddBlockRequest) (*metastorev1.AddBlockResponse, error)
+}
+
+type Recovery struct {
+	cfg    RecoveryConfig
+	l      log.Logger
+	srv    LocalServer
+	bucket objstore.Bucket
+
+	started bool
+	wg      sync.WaitGroup
+	m       sync.Mutex
+	cancel  func()
+}
+
+func NewRecovery(cfg RecoveryConfig, l log.Logger, srv LocalServer, bucket objstore.Bucket) *Recovery {
+	return &Recovery{
+		cfg:    cfg,
+		l:      l,
+		srv:    srv,
+		bucket: bucket,
+	}
+}
+
+func (r *Recovery) Start() {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if r.started {
+		r.l.Log("msg", "recovery already started")
+		return
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	r.cancel = cancel
+	r.started = true
+	go r.recoverLoop(ctx)
+	r.l.Log("msg", "recovery started")
+}
+
+func (r *Recovery) Stop() {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if !r.started {
+		r.l.Log("msg", "recovery already stopped")
+		return
+	}
+	r.cancel()
+	r.wg.Wait()
+	r.started = false
+	r.l.Log("msg", "recovery stopped")
+}
+
+const pathAnon = tenant.DefaultTenantID
+const pathBlock = "block.bin"
+const pathMetaPB = "meta.pb"
+const pathDLQ = "dlq"
```

Comment:
Let's find a place for the constants (we already have them defined in other places)

Reply:
created a separate tiny package